### PR TITLE
DevWs for NIE-TECH CO., LTD. containing containing Min Smart Plug Dimmer

### PR DESCRIPTION
**Pull request by sky-nie**

### CODE REVIEW

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy [hunk 1]
```diff
@@ -0,0 +1,672 @@
...
+       //     mnmn: "SmartThings",
+       //     vid:"generic-dimmer-3",
+            ocfDeviceType: "oic.d.smartplug"
+    ) {
+            capability "Actuator"
+            capability "Sensor"
+            capability "Switch"
+            capability "Temperature Measurement"
+            capability "Switch Level"
+            capability "Light"
```
Comment (PKacprowiczS) [inline devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: This capability is outdated

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy [hunk 2]
```diff
@@ -0,0 +1,672 @@
...
+			state "firmwareVersion", label:'Firmware ${currentValue}'
+		}
+		valueTile("icon", "device.icon", inactiveLabel: false, decoration: "flat", width: 4, height: 1) {
+			state "default", label: '', icon: "https://inovelli.com/wp-content/uploads/Device-Handler/Inovelli-Device-Handler-Logo.png"
+		}
+		main "switch", "temperature"
+		details(["switch", "temperature", "refresh", "syncStatus", "sync", "firmwareVersion"])
+	}
+
+    simulator { }
```
Comment (PKacprowiczS) [inline devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: This block isn't necessary.

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy [hunk 3]
```diff
@@ -0,0 +1,672 @@
...
+ *
+ */
+
+metadata {
+    definition (
+            name: "Min Smart Plug Dimmer",
+            namespace: "sky-nie",
+            author: "winnie",
+       //     mnmn: "SmartThings",
+       //     vid:"generic-dimmer-3",
```
Comment (PKacprowiczS) [inline devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: If you don't need `mnmn` and `vid` to be defined, just remove it.

**Review by PKacprowiczS — CHANGES_REQUESTED** [review]

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy [hunk 4]
```diff
@@ -0,0 +1,672 @@
...
+        runIn(5, executeConfigureCmds, [overwrite: true])
+    }
+
+    return []
+}
+
+private initialize() {
+    def checkInterval = ((60 * 60 * 3) + (5 * 60))
+
+    def checkIntervalEvt = [name: "checkInterval", value: checkInterval, displayed: false, data: [protocol: "zwave", hubHardwareId: device.hub.hardwareID, offlinePingable: "1"]]
```
Comment (PKacprowiczS) [inline devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: In https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/69768/files#diff-ede2fdd21911fde9f7d3dd84c6f9f40dc678f9b3f53f81350d4f6a9c30230c5fR234 you wrote that this is _sleepy_ device, but here you've marked it as `offlinePingable`, which is definitely not true for sleepy devices. Aside from the fact that this option is reserved only for crucial devices (like home security), Dimmer Plug device doesn't seem to be a sleepy device.

So, to summarize, please remove `offlinePingable: "1"`, and...

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy [hunk 5]
```diff
@@ -0,0 +1,672 @@
...
+    }
+    sendHubCommand(actions, 100)
+    return []
+}
+
+
+// Required for HealthCheck Capability, but doesn't actually do anything because this device sleeps.
+def ping() {
+	logDebug "ping()"	
+}
```
Comment (PKacprowiczS) [inline devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: ...implement proper `ping()` method which will be able to bring device back from offline state. (In other words, which will send command to which device will response, and an event will be created)

**Review by PKacprowiczS — CHANGES_REQUESTED** [review]

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy (hunk 1 above)
Comment (sky-nie) [reply devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: @PKacprowiczS :
Do you mean that the code from line 54 to line 57 is out of date? Or just line 57? My current modification is to remove the 57th line.

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy (hunk 2 above)
Comment (sky-nie) [reply devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: OK, I remove the line 117.

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy (hunk 3 above)
Comment (sky-nie) [reply devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: OK, I removed it.

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy (hunk 4 above)
Comment (sky-nie) [reply devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: OK, I removed it.

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy (hunk 5 above)
Comment (sky-nie) [reply devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: OK, I removed the line 236.

**Comment by sky-nie** [comment]
@PKacprowiczS :
I have updated the branch "devws-for-nie-tech-co-ltd-contai_49", You can review at any time convenient for you, thank you!

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy (hunk 1 above)
Comment (PKacprowiczS) [reply devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: Yeah, just `Light` capability.

**Comment by PKacprowiczS** [comment]
Answer me this question: is this really a sleepy device? Hard to believe that Dimmer Switch is actually sleepy.

**Comment by sky-nie** [comment]
> Answer me this question: is this really a sleepy device? Hard to believe that Dimmer Switch is actually sleepy.

MP21ZD And  MP22ZD with this DTH, are devices designed with micro battery. They will not go to sleep.

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy [hunk 6]
```diff
@@ -0,0 +1,656 @@
...
+    state.lastBattery = null
+    if (!state.refreshSensors) {
+        state.refreshSensors = true
+    }
+    else {
+        state.refreshConfig = true
+    }
+    refreshSyncStatus()
+    return []
+}
```
Comment (PKacprowiczS) [inline devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: Since you've confirmed that it isn't a sleepy device, can you remove any misleading code/comments stating otherwise? Also include some commands in _ping()_ command.

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy [hunk 7]
```diff
@@ -0,0 +1,656 @@
...
+
+    if (unit) {
+        eventMap.unit = unit
+        eventMap.descriptionText = "${eventMap.descriptionText}${unit}"
+    }
+    if (displayed) {
+        logDebug "${eventMap.descriptionText}"
+    }
+    return eventMap
+}
```
Comment (PKacprowiczS) [inline devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: Implementing such method and using it once seems like an overkill. I guess you can just remove it.

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy [hunk 8]
```diff
@@ -0,0 +1,656 @@
...
+        }
+        if (unit) {
+            evt.unit = unit
+        }
+        sendEvent(evt)
+    }
+    else {
+        logTrace(desc)
+    }
+}
```
Comment (PKacprowiczS) [inline devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: This will be checked by Device API, remove this method and just send an event where you're calling it.

**Review by PKacprowiczS — CHANGES_REQUESTED** [review]

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy (hunk 8 above)
Comment (sky-nie) [reply devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: I replace it with sendEvent.

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy (hunk 7 above)
Comment (sky-nie) [reply devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy:354]: I delete code as you list, I'm not sure others.

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy (hunk 6 above)
Comment (sky-nie) [reply devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: I delete code as you list, I'm not sure others.

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy (hunk 7 above)
Comment (sky-nie) [reply devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: Yes，you are right ！

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy [hunk 9]
```diff
@@ -0,0 +1,507 @@
...
+        }
+    }
+
+    state.resyncAll = false
+    def cmds1 =delayBetween(cmds, 500)
+    def actions = []
+    cmds1.each {
+        actions << new physicalgraph.device.HubAction(it)
+    }
+    sendHubCommand(actions, 100)
```
Comment (PKacprowiczS) [inline devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: I think that 
```
sendHubCommands(cmds, 500)
```
would be sufficient here, but please verify it (and change it, if that's true).

**Review by PKacprowiczS — CHANGES_REQUESTED** [review]
I think that's the last thing from me. Please check it, and we can proceed further, I suppose.

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy (hunk 9 above)
Comment (sky-nie) [reply devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: The function "sendHubCommands" is a custom function and is only called once.  I just copied the function body to the place of call.

Comment (PKacprowiczS) [reply devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: `sendHubCommand` isn't only a custom function - it is available for you right away. 
If you call it like this
```
def cmds = [List, of, commands]
sendHubCommand(cmds, delayBetweenCommands)
```
it'll be sufficient. You don't even have to call `physicalgraph.device.HubAction(cmd)`.

**Review by PKacprowiczS — CHANGES_REQUESTED** [review]
Just referring to one of my comments to get your attention https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/69768#discussion_r669750343

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy (hunk 9 above)
Comment (sky-nie) [reply devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: OK, I have changed it as you said.

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy [hunk 10]
```diff
@@ -140,17 +145,9 @@ def executeConfigureCmds() {
...
-        sendCommands(delayBetween(cmds, 500))
-    }
-    return []
-}
-
-private sendCommands(cmds) {
-    def actions = []
-    cmds?.each {
-        actions << new physicalgraph.device.HubAction(it)
+        def commands = [List, of, cmds]
```
Comment (PKacprowiczS) [inline devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: What I wrote earlier https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/69768#discussion_r669750343 was an example. It meant that object `cmds` is literally a "List of commands". Not a list containing variables `List` (which, as a side note, is a type, not a variable), `of` (which isn't even declared*) and `commands`.
In your code, you'd just use 
```
sendHubCommand(cmds, 500)
```

\* \- which additionally proves me, that you don't test your code with a device (cause you'd get an exception). Keep in mind that is your responsibility to make sure that device works correctly under your code, and it's your job to maintain it later.

**Review by PKacprowiczS — CHANGES_REQUESTED** [review]

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy (hunk 10 above)
Comment (sky-nie) [reply devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: Oh, A super low-level, stupid error...

**Review by PKacprowiczS — DISMISSED** [review]
@greens

File: devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy [hunk 11]
```diff
@@ -0,0 +1,496 @@
...
+        def storedVal = getParamStoredValue(param.num)
+        def paramVal = param.value
+        if (state.resyncAll || ("${storedVal}" != "${paramVal}")) {
+            cmds << secureCmd(zwave.configurationV1.configurationSet(parameterNumber: param.num, size: param.size, scaledConfigurationValue: paramVal))
+            if (param.num == temperatureReportThresholdParam.num) {
+                cmds << "delay 3000"
+                cmds << secureCmd(zwave.sensorMultilevelV5.sensorMultilevelGet(scale: 0, sensorType: 1))
+            } else {
+                cmds << secureCmd(zwave.configurationV1.configurationGet(parameterNumber: param.num))
+            }
```
Comment (greens) [inline devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: because of this code, your temperature report threshold will never be marked as synced

Comment (sky-nie) [reply devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: oh, I checked the history version, the bug is happened for my sync mistake.

Comment (sky-nie) [reply devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: oh, I checked the history version, the bug is happened for my sync mistake.

Comment (sky-nie) [reply devicetypes/sky-nie/min-smart-plug-dimmer.src/min-smart-plug-dimmer.groovy]: @KKlimczukS @greens I have updated [the PR](https://github.com/sky-nie/SmartThingsPublic/tree/devws-for-nie-tech-co-ltd-contai_48). Can you review it at a convenient time?

**Review by greens — APPROVED** [review]

**Comment by greens** [comment]
Should be live in production next Tuesday.
